core/client/fs/src/main/java/alluxio/client/block/policy/ConsistentHashPolicy.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.policy;
+
+import static com.google.common.hash.Hashing.murmur3_32_fixed;
+import static java.lang.Math.ceil;
+import static java.lang.String.format;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import alluxio.client.block.BlockWorkerInfo;
+import alluxio.client.block.policy.options.GetWorkerOptions;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.wire.WorkerNetAddress;
+
+import com.google.common.base.Objects;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.hash.HashFunction;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.TreeMap;
+import javax.annotation.Nullable;
+
+/**
+ * A policy based on the consistent hashing function to minimize the impact of scaling up and down.
+ */
+public class ConsistentHashPolicy implements BlockLocationPolicy {
+  private static final ConsistentHashProvider HASH_PROVIDER = new ConsistentHashProvider();
+  private final int mNumVirtualNodes;
+
+  /**
+   * Constructs a new {@link ConsistentHashPolicy}.
+   *
+   * @param conf Alluxio configuration
+   */
+  public ConsistentHashPolicy(AlluxioConfiguration conf) {
+    mNumVirtualNodes =
+        conf.getInt(PropertyKey.USER_UFS_BLOCK_READ_LOCATION_POLICY_CONSISTENT_HASH_VIRTUAL_NODES);
+  }
+
+  @Nullable
+  @Override
+  public WorkerNetAddress getWorker(GetWorkerOptions options) {
+    if (options.getBlockWorkerInfos().size() == 0) {
+      return null;
+    }
+    HASH_PROVIDER.refresh(options.getBlockWorkerInfos(), mNumVirtualNodes);
+    return HASH_PROVIDER.get(String.valueOf(options.getBlockInfo().getBlockId()), 0)
+        .getNetAddress();
+  }
+
+  @Override
+  public Set<BlockWorkerInfo> getWorkers(GetWorkerOptions options, int count) {
```

Comment:
In the original logic in Presto this is a list and the workers are ordered. The implementation here is different and might affect performance.